Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java

```diff
@@ -368,8 +371,154 @@ private void updateDriverStatus(Boolean status) {
+                }
+                JSONObject res = (JSONObject) new JSONArray(String.valueOf(result)).get(0);
+                JSONArray pathLatLng = (JSONArray) res.get("points");
+
+                SharedPreferences sharedPreff = this.getSharedPreferences(this.getString(R.string.preference_file_key), Context.MODE_PRIVATE);
+                SharedPreferences.Editor editor = sharedPreff.edit();
+                editor.putString("RIDE_WAYPOINT", String.valueOf(pathLatLng));
+                editor.apply();
+            }
+
+        } catch (Exception e) {
+            Log.e("get route api exception", e.toString());
+        }
+    }
+
+    private void updateDeviation(double latitude, double longitude, String waypoints) throws JSONException {
+        try {
+
+            SharedPreferences sharedPreff = this.getSharedPreferences(this.getString(R.string.preference_file_key), Context.MODE_PRIVATE);
+            List<LatLng> LatLngPoints = new ArrayList<LatLng>();
+            JSONArray waypointArray = new JSONArray(waypoints);
+
+            for (int i = 0; i < waypointArray.length(); i++) {
+                JSONObject pt = waypointArray.getJSONObject(i);
+                LatLngPoints.add(new LatLng((Double) pt.get("lat"), (Double) pt.get("lon")));
+            }
+
+
+            int resultIndex = PolyUtil.locationIndexOnEdgeOrPath(new LatLng(latitude, longitude), LatLngPoints, PolyUtil.isClosedPolygon(LatLngPoints), true, 30.0);
+
+            if (resultIndex == -1) {
+                int deviation = Integer.parseInt(sharedPreff.getString("RIDE_WAYPOINT_DEVIATION_COUNT", null));
+                deviation = deviation + 1;
+                SharedPreferences.Editor editor = sharedPreff.edit();
+                editor.putString("RIDE_WAYPOINT_DEVIATION_COUNT", String.valueOf(deviation));
+                editor.apply();
+
+                Log.d("new deviation count" , String.valueOf(deviation));
+            } else {
+               Log.d(  "in path", String.valueOf(resultIndex));
+            }
+        }
+        catch (Exception e) {
+            Log.e("deviation update exception", e.toString());
+        }
+    }
+
     /*Location update API call*/
-    private void callDriverCurrentLocationAPI(double latitude, double longitude, float accuracy, String locTime, String log, String locationSource, String triggerFunctionValue) {
+    private void callDriverCurrentLocationAPI(double latitude, double longitude, float accuracy, String locTime, String log, String locationSource, String triggerFunctionValue) throws JSONException {
+
+        try {
+            SharedPreferences sharedPreff = this.getSharedPreferences(this.getString(R.string.preference_file_key), Context.MODE_PRIVATE);
+            String localStage = sharedPreff.getString("LOCAL_STAGE", null);
+
+            if (localStage.equals("RideStarted")) {
+                String waypoints = sharedPreff.getString("RIDE_WAYPOINT", null);
+
+                if (waypoints.isEmpty()) {
```

Comment:
isEmpty() will give NullPointerException if applied on a null value. Which is possible as the default value passed while getting it from storage is `null`. 

You can have checks in two ways:-

`Objects.isNull(waypoints)`

or 

`waypoints == null`